src/Swarm/Language/Typecheck.hs

```diff
@@ -218,10 +219,12 @@ generalize uty = do
   ctxfvs <- freeVars ctx
   let fvs = S.toList $ tmfvs \\ ctxfvs
       alphabet = ['a' .. 'z']
-      xs =
-        take (length fvs) $
-          map T.pack (map (: []) alphabet ++ [x : show n | n <- [0 :: Int ..], x <- alphabet])
-  return $ Forall xs (substU (M.fromList (zip (map Right fvs) (map UTyVar xs))) uty')
+      -- infinite supply of pretty names a, b, ..., z, a0, ... z0, a1, ... z1, ...
+      prettyNames = map T.pack (map (: []) alphabet ++ [x : show n | n <- [0 :: Int ..], x <- alphabet])
+  return $
+    Forall
+      (zipWith const prettyNames fvs)   -- one pretty name per free variable
+      (substU (M.fromList . map (Right *** UTyVar) $ zip fvs prettyNames) uty')
```

Comment:
@kostmo is this more readable?

Reply:
I like that `prettyNames` is an infinite list now.  But I wonder if there is a readable way to avoid referencing it+zipping with it twice on lines 226-227 :smile: ?